fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java

```diff
@@ -319,65 +322,40 @@ public OptExpression visitPhysicalOlapScan(OptExpression optExpression, DecodeCo
-                                        couldApplyDictOptimize(predicate));
-                        if (!couldApply) {
-                            globalDictStringColumns.remove(stringColumn);
-                            dictStringIdToIntIds.remove(stringColumn.getId());
-                            couldEncoded = false;
-                        } else {
-                            for (int i = 0; i < predicates.size(); i++) {
-                                ScalarOperator predicate = predicates.get(i);
-                                if (predicate.getUsedColumns().contains(columnId)) {
-                                    Preconditions.checkState(couldApplyDictOptimize(predicate));
-                                    if (newDictColumn == null) {
-                                        newDictColumn = context.columnRefFactory.create(
-                                                stringColumn.getName(), ID_TYPE, stringColumn.isNullable());
-                                    }
-
-                                    final DictMappingOperator newCallOperator =
-                                            new DictMappingOperator(newDictColumn, predicate.clone(),
-                                                    predicate.getType());
-
-                                    predicates.set(i, newCallOperator);
-                                }
-                            }
+                    BooleanSupplier checkColumnCouldApply = () -> {
+                        if (context.disableDictOptimizeColumns.contains(columnId)) {
+                            return false;
                         }
-                    }
 
-                    if (!couldEncoded) {
+                        if (scanOperator.getPredicate() != null &&
+                                scanOperator.getPredicate().getUsedColumns().contains(columnId)) {
+                            // If there is an unsupported expression in any of the low cardinality columns,
+                            // we disable low cardinality optimization.
+                            return predicates.stream()
+                                    .allMatch(predicate -> !predicate.getUsedColumns().contains(columnId) ||
+                                            couldApplyDictOptimize(predicate, context.allStringColumnIds));
+                        }
+                        return true;
+                    };
+
+                    if (!checkColumnCouldApply.getAsBoolean()) {
+                        context.allStringColumnIds.remove(columnId);
                         continue;
                     }
 
-                    if (newDictColumn == null) {
-                        newDictColumn = context.columnRefFactory.create(
-                                stringColumn.getName(), ID_TYPE, stringColumn.isNullable());
-                    }
+                    ColumnRefOperator newDictColumn = context.columnRefFactory.create(
```

Comment:
suggest to add a function has a meaningful name to enclose this code, then invoke this function. I think this function
would be invoked for many times, advice to use the name `substituteStringColumnWithDictEncodedColumn`.